### PR TITLE
Use of std::vector and other locals in parallel_for

### PR DESCRIPTION
**Pull request by bathmatt**
@trilinos/panzer

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ ] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp [hunk 1]
```diff
@@ -137,18 +137,20 @@ namespace panzer
...
-    for (const auto& name : resNames)
+    fields_ = Kokkos::View<PHX::MDField<ScalarT, Cell, BASIS>*>("Integrator_GradBasisCrossVector::fields_", resNames.size());
+
     {
-      MDField<ScalarT, Cell, BASIS> res(name, basis.functional);
-      fields_.push_back(res);
+      int i=0;
+      for (const auto& name : resNames)
+        fields_(i++) = MDField<ScalarT, Cell, BASIS>(name, basis.functional);
     } // end loop over resNames
```
Comment (jmgate) [inline packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp]: I'm not understanding what this block is doing without the `for` at the top of it.  Does the whole block need to be removed?

File: packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp [hunk 2]
```diff
@@ -137,18 +137,20 @@ namespace panzer
...
-    for (const auto& field : fields_)
+    for (std::size_t i=0; i< fields_.extent(0); ++i) {
+      const auto& field = fields_(i);
       if (evalStyle_ == EvaluatorStyle::CONTRIBUTES)
         this->addContributedField(field);
       else // if (evalStyle_ == EvaluatorStyle::EVALUATES)
         this->addEvaluatedField(field);
-
+    }
```
Comment (jmgate) [inline packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp]: I think you probably want the `if` outside the `for`.

File: packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisTimesScalar_impl.hpp [hunk 3]
```diff
@@ -119,17 +119,19 @@ namespace panzer
...
+      for (const auto& name : resNames)
+        fields_(i++) = MDField<ScalarT, Cell, BASIS>(name, basis.functional);
+    }
+    for (std::size_t i=0; i<fields_.extent(0); ++i) {
+      const auto& field = fields_(i);
       if (evalStyle_ == EvaluatorStyle::CONTRIBUTES)
         this->addContributedField(field);
       else // if (evalStyle_ == EvaluatorStyle::EVALUATES)
         this->addEvaluatedField(field);
+    }
```
Comment (jmgate) [inline packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisTimesScalar_impl.hpp]: See comments on the previous file.

File: packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp (hunk 1 above)
Comment (bathmatt) [reply packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp]: It's assigning the values to fields_...
I guess I don't understand your question.  It's a loop over resNames that allocs data and assigns it

Comment (jmgate) [reply packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp]: Nevermind&mdash;it was just the way the diff was rendered that confused me.  If I were you, I'd remove the surrounding `{}` and outdent lines 143&ndash;145.

File: packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp (hunk 2 above)
Comment (jmgate) [reply packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_GradBasisCrossVector_impl.hpp]: Do you want to switch this up such that we're not doing an `if` every iteration?  If not, I'll dismiss my review and you can merge.

**Pull request merged**